doc/kernel/usermode/usermode.rst

```diff
@@ -9,6 +9,176 @@ are defined, and how memory may be managed to support user mode threads.
+  with their own group of threads and private data structures, which are
+  isolated from each other if one crashes or is otherwise compromised.
+
+Design Goals
+------------
+
+For threads running in a non-privileged CPU state (hereafter referred to as
+'user mode') we aim to protect against the following:
+
+- We prevent access to memory not specifically granted, or incorrect access to
+  memory that has an incompatible policy, such as attempting to write to a
+  read-only area.
+
+  - Threads are automatically granted access to their own stack memory
+    region, and all other stacks are inaccessible.
+
+  - By default, program text and read-only data are accessible to all threads
+    on read-only basis, kernel-wide. This policy may be adjusted.
+
+  - If the optional "application memory" feature is enabled, then all
+    non-kernel globals defined in the application and libraries will be
+    accessible.
+
+- We prevent use of device drivers or kernel objects not specifically granted,
+  with the permission granularity on a per object or per driver instance
+  basis.
+
+- We validate kernel or driver API calls with incorrect parameters that would
+  otherwise cause a crash or corruption of data structures private to the
+  kernel. This includes:
+
+  - Using the wrong kernel object type.
+
+  - Using parameters outside of proper bounds or with nonsensical values.
+
+  - Passing memory buffers that the calling thread does not have sufficient
+    access to read or write, depending on the semantics of the API.
+
+  - Use of kernel objects that are not in a proper initialization state.
+
+- We ensure the detection and safe handling of user mode stack overflows.
+
+- We prevent invoking system calls to functions excluded by the kernel
+  configuration.
+
+- We prevent disabling of or tampering with kernel-defined and hardware-
+  enforced memory protections.
+
+- We prevent re-entry from user to supervisor mode except through the kernel-
+  defined system calls and interrupt handlers.
+
+- We prevent the introduction of new executable code by user mode threads,
+  except to the extent to which this is supported by kernel system calls.
+
+We are specifically not protecting against the following attacks:
+
+- The kernel itself, and any threads that are executing in supervisor mode,
+  are assumed to be trusted.
+
+- The kernel build is assumed to be trusted. There is considerable build-time
+  logic for creating the tables of valid kernel objects, defining system calls,
+  and configuring interrupts. The .elf binary files that are worked with
+  during this process are all assumed to be trusted code.
+
+- We can't protect against mistakes made in memory domain configuration done in
+  kernel mode that exposes private kernel data structures to a user thread. RAM
+  for kernel objects should always be configured as supervisor-only.
+
+- It is possible to make top-level declarations of user mode threads and
```

Comment:
Maybe mention that the compilers and tools supplied with the Zephyr SDK are also assumed to be trusted?

Reply:
good idea.